File: backend/models/Booking.js
const mongoose = require('mongoose');

const passengerSchema = new mongoose.Schema({
  name: { type: String, required: true },
  age: { type: String, required: true },
  gender: { type: String, required: true },
  berth_preference: { type: String, required: true },
  coach: { type: String, required: true },
  seat_number: { type: String, required: true }
});

const bookingSchema = new mongoose.Schema({
  pnr: { type: String, required: true, unique: true },
  train_number: { type: String, required: true },
  train_name: { type: String, required: true },
  source: { type: String, required: true },
  destination: { type: String, required: true },
  departure_time: { type: String, required: true },
  arrival_time: { type: String, required: true },
  journey_date: { type: String, required: true },
  status: { type: String, required: true, default: 'CONFIRMED' },
  total_amount: { type: Number, required: true },
  payment_mode: { type: String, required: true },
  transaction_id: { type: String, required: true },
  user_email: { type: String, required: true, lowercase: true },
  booking_date: { type: Date, default: Date.now },
  passengers: [passengerSchema]
});

module.exports = mongoose.model('Booking', bookingSchema);

File: lib/models/booking.dart
import 'train.dart';
import 'passenger.dart';

class Booking {
  final String pnr;
  final Train train;
  final String journeyDate;
  final List<Passenger> passengers;
  String status; // 'CONFIRMED', 'CANCELLED'
  final double totalAmount;
  final String paymentMode;
  final String transactionId;
  final DateTime bookingDate;
  final String? bookedSource;
  final String? bookedDestination;

  Booking({
    required this.pnr,
    required this.train,
    required this.journeyDate,
    required this.passengers,
    required this.status,
    required this.totalAmount,
    required this.paymentMode,
    required this.transactionId,
    required this.bookingDate,
    this.bookedSource,
    this.bookedDestination,
  });

  Map<String, dynamic> toJson() => {
        'pnr': pnr,
        'train': train.toJson(),
        'journeyDate': journeyDate,
        'passengers': passengers.map((p) => p.toJson()).toList(),
        'status': status,
        'totalAmount': totalAmount,
        'paymentMode': paymentMode,
        'transactionId': transactionId,
        'bookingDate': bookingDate.toIso8601String(),
        'bookedSource': bookedSource,
        'bookedDestination': bookedDestination,
      };

  factory Booking.fromJson(Map<String, dynamic> json) => Booking(
        pnr: json['pnr'],
        train: Train.fromJson(json['train']),
        journeyDate: json['journeyDate'],
        passengers: (json['passengers'] as List)
            .map((p) => Passenger.fromJson(p))
            .toList(),
        status: json['status'],
        totalAmount: json['totalAmount'],
        paymentMode: json['paymentMode'],
        transactionId: json['transactionId'],
        bookingDate: DateTime.parse(json['bookingDate']),
        bookedSource: json['bookedSource'],
        bookedDestination: json['bookedDestination'],
      );
}
